rust/kernel/str.rs

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! String representations.
+
+use core::ops::{self, Deref, Index};
+
+use crate::bindings;
+use crate::c_types;
+
+/// Byte string without UTF-8 validity guarantee.
+///
+/// `BStr` is simply an alias to `[u8]`, but has a more evident semantical meaning.
+pub type BStr = [u8];
+
+/// Creates a new [`BStr`] from a string literal.
+///
+/// `b_str!` converts the supplied string literal to byte string, so non-ASCII
+/// characters can be included.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// const MY_BSTR: &'static BStr = b_str!("My awesome BStr!");
+/// ```
+#[macro_export]
+macro_rules! b_str {
+    ($str:literal) => {{
+        const S: &'static str = $str;
+        const C: &'static $crate::str::BStr = S.as_bytes();
+        C
+    }};
+}
```

Comment:
Isn't this just `b"foo"`?

Reply:
I think the point is to accept non-ascii characters:

```
   |
32 |     let t = b"ão";
   |               ^
   |               |
   |               byte constant must be ASCII
   |               help: use a \xHH escape for a non-ASCII byte: `\xE3`
```

Reply:
Ah, ok.